### DISCUSSION TITLE
Welcome to parallel-gpu-workshop-JuliaCon21 Discussions!

### DISCUSSION THREAD

**Discussion by luraess** [luraess/parallel-gpu-workshop-JuliaCon21#1] [General]
## 👋 Welcome!
  We’re using Discussions as a place to exchange about topics related to this short course. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Welcome others, stay open-minded and respectful.

**Comment by chris-revell** [comment]
Hello. Thank you for presenting this workshop at JuliaCon; it has been very helpful. I wonder if you could explain or provide a reference for how the dtau time step is set within the psuedotime iteration for the implicit methods? I can see that is is calculated in the following line: `dtau   .= (1.0./(min(dx, dy)^2 ./inn(H).^npow./4.1) .+ 1.0/dt).^-1`. However, it is not immediately obvious to me how this expression was derived from the basic CFL condition, or how it should be adapted for other PDEs. Thanks very much.

**Reply by luraess** [reply]
Hi @chris-revell, thanks for your feedback!

The definition of `dtau` builds upon two parts, (1) the diffusion CFL-like definition (PDE) and (2) the `dt` limiter for the temporal discretisation of the physical time-step (ODE):
- In (1), we make analogy to linear diffusion where the stable time step definition would be the CFL one, namely `dt_diff = min(dx,dy)^2/D/4.1` in 2D. In our case, the effective diffusion coefficient is approximated such that `D = H^npow`.
- The ODE term for time integration limiter is (2) `dt`.

Then we construct the final effective iterative pseudo-time step `dtau` as the harmonic mean of `dt_diff` and `1.0/dt`, such that `dtau = (1.0/dt_diff + 1.0/dt)^-1` in order to get a smooth transition between the two conditions and ensure to use the most restrictive one. I am right now preparing a paper to be certainly submitted to _Journal of Computational Physics_ on the derivation and determination of pseudo-time stepper for the implicit iterative pseudo-transient method.

**Reply by chris-revell** [reply]
Thanks for getting back to me. Part (1) is what I found perplexing. In your bullet point you describe how the stability condition for linear diffusion is `dt_diff = min(dx,dy)^2/D/4.1`. Is this empirical or derived from somewhere? Basically, I'm unsure how to set this pseudo time step for the particular equation that I want to solve, hence I'm keen to understand where the time step that you have used has come from.

**Reply by luraess** [reply]
> Is this empirical or derived from somewhere?

The definition `dt_diff = min(dx,dy)^2/D/4.1` is the CFL-type condition to have conditionally stable explicit scheme. The 1D analogy reads `∆t ≤ ∆x^2/D/2` and comes from the fact that the amount added to the explicit update of the quantity that diffuses, e.g. `u`, should ≤ 1. In `u[ix,it+1] = u[ix,it] + ∆t/∆x^2*D*(u[ix-1,it]-2*u[ix,it]+u[ix+1,it])`, the update term `u[ix,it]` has `∆t/∆x^2*D*2` as factor, thus if one want `∆t/∆x^2*D*2 ≤ 1`, i.e. `∆t ≤ ∆x^2/D/2` or `∆t = D*∆x/2.1`.

**Reply by chris-revell** [reply]
I see. Sorry to keep following up with more questions, but why is the `u[ix,it]` term the one from which you take the prefactor, rather than the full expression for the flux?

**Reply by luraess** [reply]
Formally, things may get slightly more complex. The explanation from this thread should help for basic understanding and building intuition. If interested to further dig into explicit time step definition, I would recommend to familiarize with stability analysis procedures.

Besides the CFL condition, the stability condition needs also to be ensured in order to have scheme converging to the correct solution; an example of unconditionally stable but sometimes unphysical scheme is a Crank-Nicolson scheme for diffusion equation (https://en.wikipedia.org/wiki/Crank-Nicolson_method).

A generic procedure to obtain such stability conditions is e.g. the von Neumann stability analysis (https://en.wikipedia.org/wiki/Von_Neumann_stability_analysis); it ensures that the addition of a small perturbation to the round-off error does not grow infinitely in time. The Wikipedia page introduces the stability analysis exactly for the diffusion equation. Adding to the RHS of the diffusion equation the `u/dt` term will appear in the final answer exactly as in the formula in the codes.

**Reply by chris-revell** [reply]
Thanks for the detailed response. Seems like I've got a lot to look into here. Some literature suggests that a split implicit/explicit scheme may be best for the equation I want to solve (conserved Swift-Hohenberg), which would avoid the pseudo time iteration, so I might pursue that avenue. The GPU functionality remains the primary reason I was interested in this tutorial, the time stepping was just an interesting bonus!

**Reply by luraess** [reply]
Thanks for feedback! Seems the cSH equation is a cubic nonlinear one. You could give it a try with the pseudo-transient iterative approach as then you could leverage GPU performance for stencil computations (where GPUs are good at). Implicit/explicit schemes may work also work - question is which part is implicit and what does implicit mean there.

**Reply by chris-revell** [reply]
Ah, so is the pseudo-transient iterative approach a requirement to best leverage the GPU? I was looking at the split scheme described here https://www.sciencedirect.com/science/article/pii/S0021999116001947, which, to my current poor understanding, involves an approximation to linearise one component, which can be solved implicitly, leaving another component to be solved explicitly.

**Reply by luraess** [reply]
> pseudo-transient iterative approach a requirement to best leverage the GPU

No, don't get me wrong - sorry if I was unclear. I was just pointing out that if you use an (partly) implicit scheme, then you are likely to need the solution of a linear system of equation (in your ref a pseudo-spectral approach seems to be used). This may be the performance bottleneck of your application as approximating the inverse of (large) sparse arrays may become costly in terms of memory pressure.

**Reply by chris-revell** [reply]
That makes sense. Would you say that the GPU is less effective for schemes that require matrix inversion?

**Reply by luraess** [reply]
> the GPU is less effective for schemes that require matrix inversion

Not exactly. There may be a trade-off. The GPU can certainly handle "matrix inversions" well, but the implementation may not scale; the larger the problem size, the nonlinearly more memory transfers may be needed which may turn out to be the bottleneck at some point.

**Reply by chris-revell** [reply]
Thanks for the input. Seems like I've got a lot to look into!
